### ISSUE TITLE
Propose and implement synchronization flow for multicore power management

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1003]
**_Reported by Julien Delayen:_**

In order to synchronize multiple cores to enter sleep, a synchronization mechanism
is needed.
The synchronization will be used to ensure that both cores are able to enter sleep
(ie: not busy) and perform the setup needed before actually performing the transition
into sleep.
The x86 or sensor core will send a request to enter sleep, the core sending the request
will perform the final step to transition to sleep mode.
In case of one core not being able to enter sleep, the other core will enter a low
power state and will be able to resume to an active state by discarding the setup
previously performed. 
This mechanism could be using some channels from the ipm driver.

(Imported from Jira ZEP-1103)

**Comment by nashif** [comment]
**_by Andre Guedes:_**

Me, Ramesh, Ivan, and Julien have discussed and come up with a proposal on how to handle the PM multicore scenario for Quark SE.

The proposal is based on a master/slave approach where the x86 core always plays the 'master' role while the ARC core plays the 'slave' role. Only the 'master' can put the system in sleep mode.

In order to keep the implementation as simple as possible, we should use a bit flag in GP0 register (which is shared by both cores) instead of mailbox.

The approach works as follows: ARC PMA saves the context of all peripherals ARC is using and put the system in LPS state. Before entering in SS2, ARC power driver saves the execution context, calls power_soc_set_ss_restore_flag(), and raise a bit flag in GP0 register to indicate that ARC is halted and it is ready to enter in DEEP_SLEEP, in case x86 core decides to.

When x86 PMA requests to enter in DEEP_SLEEP state, the power driver checks the flag in GP0. If it is not set (which means ARC is busy), it returns error to PMA. Otherwise, it saves x86 context and puts the system in DEEP_SLEEP state.

Finally, when the wake up event is fired, the x86 core restores its context and restart ARC which also restores its context.

**Comment by nashif** [comment]
**_by Ramesh Thomas:_**

Suggest create DEEP_SLEEP_2 for ARC and LMT for multi-core deep sleep purpose.

There was a race condition discussed, when ARC can be woken up while LMT is in the process of entering deep sleep.  Suggest the following sequence to handle it. Not doing so will cause system to be put to deep sleep when ARC is in an unknown state.

```
ARC                                   LMT
====                                  ====
_sys_soc_suspend:           _sys_soc_suspend:
                                          flag_lmt = 1
save context
flag_arc = 1                       if flag_arc == 1
enter LPS                              save context
                                              flag_arc = 0
                                              Send IPM or trigger any event that can wake ARC
                                              while (flag_arc == 0 && time < arc_ready_wait_threshold_time) 
                                                  nop
                                              if (flag_arc == 1)
                                                  enter deep sleep
post_ops:
     if flag_lmt == 1
          save context
          flag_arc = 1
          while (flag_lmt == 1)     
     flag_arc = 0

```

**Comment by nashif** [comment]
**_by Sharron LIU:_**

Please update sample power app to demonstrate the usage of the proposal. Thanks.

**Comment by nashif** [comment]
**_by Andre Guedes:_**

Yes, sure. I'll update sample/power/power_mgr or provide a new sample apps to test the multicore scenario.

**Comment by nashif** [comment]
**_by Andre Guedes:_**

Patch implementing the PoC submitted to gerrit and reviewed.

The patch also adds the sample application test/power/quark_se/ which provides an easy way to verify the feature. Instructions on how to build and run the sample application is provided in test/power/quark_se/README file.

The suggested code doesn't seem to handle the corner case we have been discussing properly. For instance, the wake interrupt can be fired while 'post_ops' is executing. The wake interrupt will be pending while the x86 core puts the SoC in sleep state. Since the wake interrupt was already fired, the system will sleep forever.

I also spent some time thinking about other ways to handle this corner case, but I didn't find a way. By the way, we might have a similar corner case with x86-only scenario (the wake interrupt is fired while the SoC is about to enter in sleep state). I suggest we have another story to sort this out. So that corner case is not covered by the PoC code.

**Comment by nashif** [comment]
**_by Ramesh Thomas:_**

>>the wake interrupt can be fired while 'post_ops' is executing. The wake interrupt will be pending while the x86 core puts the SoC in sleep state. Since the wake interrupt was already fired, the system will sleep forever.

post_ops is entered because the wake event at ARC already fired.  Wouldn't you be setting a new one?  In that case wouldn't it be set for a long enough interval so it does not fire immediately? 

Note that the main idea is for LMT side to deterministically cause the ARC to enter "ready for deep sleep" state.  Another way to do it is for LMT to control when the wake event happens i.e. the app should be sure there will be no wake event when LMT is entering deep sleep.

**Comment by nashif** [comment]
**_by Andre Guedes:_**

> post_ops is entered because the wake event at ARC already fired. Wouldn't you be setting a new one? In that case wouldn't it be set for a long enough interval so it does not fire immediately?

Not really. According to the algorithm suggested, post_ops is entered because x86 sent an IPM message or triggered any event. This is not the wakeup event.

> Note that the main idea is for LMT side to deterministically cause the ARC to enter "ready for deep sleep" state. Another way to do it is for LMT to control when the wake event happens i.e. the app should be sure there will be no wake event when LMT is entering deep sleep.

I don't think we can assume anything about the wakeup event e.g. the wakeup event could be an AON GPIO interrupt which can be fired any time. We have no control when it will happen.

**Comment by nashif** [comment]
**_by Ramesh Thomas:_**

>>Not really. According to the algorithm suggested, post_ops is entered because x86 sent an IPM message or triggered any event. This is not the wakeup event.

You are right. We can still setup a new wake event here.  In fact we should setup wake event only after LMT is ready.  Why not let LMT set wake event as it is the master and should be in control of when system wakes up.

>>I don't think we can assume anything about the wakeup event e.g. the wakeup event could be an AON GPIO interrupt which can be fired any time. We have no control when it will happen.

But we do have control of the actual wake event that was setup while entering deep sleep.  We have control if we set it up when LMT is entering deep sleep.

Can you also consider the method proposed in gerrit comment i.e. implement this as purely master-slave with LMT controlling when ARC enters DEEP_SLEEP_2.  That can be achieved by LMT sending IPM and ARC entering DEEP_SLEEP_2 in the handler of the IPM.

**Comment by nashif** [comment]
**_by Andre Guedes:_**

I'm afraid we cannot setup the wake event in post_ops. The user is the only one who knows what the wake event is, so the application level seems indeed the right place to setup the wake event. Moreover, the "wakeup device" is used and controlled by the ARC application. Having the LMT responsible by configuring the "wakeup device" seems a bit odd. Not sure I fully understand how this approach would fix the corner case, though. Maybe we need an execution flow example.

**Comment by nashif** [comment]
**_by Andre Guedes:_**

Besides that, we should probably figure out first how to handle this corner case in the x86-only scenario since it is simpler than the multicore core scenario. The corner case is: if the wake interrupt is fired while we are putting the system in deep sleep, the deep sleep procedure is not aborted (interrupts are masked at this point) and we will sleep forever since the wake interrupt has been already fired.

**Comment by nashif** [comment]
**_by Ramesh Thomas:_**

The corner case you are mentioning does not look like a real one. When the wake event is set, why would it be set in such a way that it could get triggered immediately?  The issue you are referring to is because the decision to set the wake event at the ARC side is now out of sync with the LMT side which is the one that enters deep sleep.

It makes sense to set or control the setting of the wake event from LMT side because you are following master-slave approach.  For this app, can't you just use a device controlled by LMT?  If you want to use ARC side wake event then LMT should be able to tell ARC to setup the wake event when it is ready to enter deep sleep.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

_Andre Guedes_ , what's the current status?  When do you expect to get the code review completed so QA can finish their testing?  As a reminder, QA's testing of the PwrMgmt features is currently gated by this story.

**Comment by nashif** [comment]
**_by Andre Guedes:_**

The patch implementing the approach Ramesh asked above was posted yesterday. Review is still going on. The review process doesn't depend on me but I hope it will finish soon.

**Comment by nashif** [comment]
**_by Sharron LIU:_**

Verified with QMSI v1.4 RC3, with patch of GH-1629.

Closing this. Thanks1
